connection.go

```diff
@@ -28,19 +28,22 @@ type mysqlContext interface {
 }
 
 type mysqlConn struct {
-	buf              buffer
-	netConn          net.Conn
-	affectedRows     uint64
-	insertId         uint64
-	cfg              *Config
-	maxAllowedPacket int
-	maxWriteSize     int
-	writeTimeout     time.Duration
-	flags            clientFlag
-	status           statusFlag
-	sequence         uint8
-	parseTime        bool
-	strict           bool
+	buf                 buffer
+	netConn             net.Conn
+	affectedRows        uint64
+	insertId            uint64
+	cfg                 *Config
+	maxAllowedPacket    int
+	maxWriteSize        int
+	writeTimeout        time.Duration
+	flags               clientFlag
+	status              statusFlag
+	sequence            uint8
+	compressionSequence uint8
+	parseTime           bool
+	strict              bool
+	reader              packetReader
+	writer              io.Writer
```

Comment:
My understanding of your comment is that you suggest merging reader and buf, and writer and netConn. Because both buf and netConn are used in various capacities outside of simple reading and writing (e.g. setting deadlines or getting the complete buffer) , it's not easily possible to subsume buf and netConn into reader and writer respectively. 
 
I wrote things thusly to avoid having the addition of compression significantly affect buffer.go. While this refactoring sounds like a great idea, I think that it may be out of the scope of this ticket, and would be better undertaken in a separate ticket with some design/brainstorming done beforehand. This would also avoid obfuscating the scope of this pull request, which was only the addition of the compression protocol. 

In re: sequence numbers: 
Because sequence and compression sequence need to be reset on the basis of higher level actions (eg. issuing a new mysql command), it seems that a writer-level abstraction should not know about specific mysql connection logic.